Data/Primitive/ByteArray.hs

```diff
@@ -181,12 +180,16 @@ foldrByteArray f z arr = go 0
     go i
       | sizeofByteArray arr > i * sz = f (indexByteArray arr i) (go (i+1))
       | otherwise                    = z
-    sz = sizeofByteArray arr
+    sz = sizeOf (undefined :: a)
 
 fromListN :: Prim a => Int -> [a] -> ByteArray
-fromListN n xs = runST $ do
-    marr <- newByteArray (n * sizeOf (head xs))
-    zipWithM_ (writeByteArray marr) [0..n] xs
+fromListN n ys = runST $ do
+    marr <- newByteArray (n * sizeOf (head ys))
+    let go !_ [] = return ()
+        go !ix (x : xs) = do
+          writeByteArray marr ix x
+          go (ix + 1) xs
+    go 0 ys
```

Comment:
While you're replacing the `IsList` instance, I think you should make it safe by checking against the given size on each iteration. I'm adding that check to the polymorphic arrays.

Reply:
I’ll do that.  It’s a pretty inexpensive check, and this function  should never be used in performance-critical situations, and segfaults  are terrible.